### DISCUSSION TITLE
Listing routes at runtime

### DISCUSSION THREAD

**Discussion by morzel85** [TanStack/router#998] [Q&A]
Such code allows me to get a TypeScript union of available routes (from chosen tree):

`export type ExampleRoutes = ParseRoute<typeof exampleTree>['fullPath'];`

Static checks, nice!

Is it also possible to get a list of routes that will be usable at runtime?
Let's say I want to iterate through routes or validate them...

Maybe its possible to get a string enum that is usable both in types and values space?

**Comment by schiller-manuel** [comment]
How about `router.routesByPath`?

**Reply by flipmotion** [reply]
Hi @schiller-manuel 

Try to check this following
```typescript
export type GetTFrom<T, TRouteTree extends AnyRoute> = T extends StrictOrFrom<infer TFrom extends RouteIds<TRouteTree>> ? TFrom : never;
```

this will return all suggestions of your router as list

I guess you can just check how was implemented `useSearch/useParams`

**Comment by seedy** [comment]
With `useRouter` hook, I found 2 different solutions, thanks to @schiller-manuel :

```
import {useRouter } from '@tanstack/react-router';


const {routesByPath, flatRoutes} = useRouter();

// routesByPath is an object with routes pathnames as keys and Route objects as values
// flatRoutes is an array of Route objects
```

**Reply by ianzone** [reply]
This should be listed in docs

**Reply by lpolon** [reply]
The initial question suggested that an easy way to get the types of all available routes is like:

```ts
export type ExampleRoutes = ParseRoute<typeof exampleTree>['fullPath'];
```

however, as far as I can tell, `ParseRoute` is not available anymore:
https://github.com/TanStack/router/discussions/824#discussioncomment-12763935

and is not documented:
https://tanstack.com/router/v1/docs/framework/react/guide/type-utilities

but given `useRouter().routesByPath`, looks straightforward to get the types like:

```ts
export type ExampleRoutes = keyof RegisteredRouter['routesByPath'];
```

**Reply by VictorPulzz** [reply]
@lpolon Thanks a lot!
